picasso/src/test/java/com/squareup/picasso/SourceMarkerTest.java

```diff
@@ -0,0 +1,251 @@
+    BufferedSource source = marker.source();
+
+    long pos0 = marker.mark(0);
+    marker.reset(pos0);
+    assertThat(source.readUtf8(3)).isEqualTo("ABC");
+  }
+
+  @Test public void markNegative() throws Exception {
+    SourceMarker marker = new SourceMarker(new Buffer().writeUtf8("ABCDEFGHIJKLMNOPQRSTUVWXYZ"));
+    BufferedSource source = marker.source();
+
+    try {
+      marker.mark(-1L);
+      fail();
+    } catch (IllegalArgumentException expected) {
+      assertThat(expected).hasMessageThat().isEqualTo("readLimit < 0: -1");
+    }
+  }
+
+  @Test public void resetAfterEof() throws Exception {
+    SourceMarker marker = new SourceMarker(new Buffer().writeUtf8("ABCDE"));
+    BufferedSource source = marker.source();
+
+    long pos0 = marker.mark(5);
+    assertThat(source.readUtf8()).isEqualTo("ABCDE");
+    marker.reset(pos0);
+    assertThat(source.readUtf8(3)).isEqualTo("ABC");
+  }
+
+  @Test public void closeThenMark() throws Exception {
+    SourceMarker marker = new SourceMarker(new Buffer().writeUtf8("ABCDEFGHIJKLMNOPQRSTUVWXYZ"));
+    BufferedSource source = marker.source();
+
+    source.close();
+    try {
+      marker.mark(5);
+      fail();
+    } catch (IllegalStateException expected) {
+      assertThat(expected).hasMessageThat().isEqualTo("closed");
+    }
+  }
+
+  @Test public void closeThenReset() throws Exception {
+    SourceMarker marker = new SourceMarker(new Buffer().writeUtf8("ABCDEFGHIJKLMNOPQRSTUVWXYZ"));
+    BufferedSource source = marker.source();
+
+    long pos0 = marker.mark(5);
+    source.close();
+    try {
+      marker.reset(pos0);
+      fail();
+    } catch (IllegalStateException expected) {
+      assertThat(expected).hasMessageThat().isEqualTo("closed");
+    }
+  }
+
+  @Test public void closeThenRead() throws Exception {
+    SourceMarker marker = new SourceMarker(new Buffer().writeUtf8("ABCDEFGHIJKLMNOPQRSTUVWXYZ"));
+    BufferedSource source = marker.source();
+
+    source.close();
+    try {
+      source.readUtf8(3);
+      fail();
+    } catch (IllegalStateException expected) {
+      assertThat(expected).hasMessageThat().isEqualTo("closed");
+    }
+  }
+
+  @Test public void multipleSegments() throws Exception {
+    String as = repeat('a', 10_000);
+    String bs = repeat('b', 10_000);
+    String cs = repeat('c', 10_000);
+    String ds = repeat('d', 10_000);
+
+    SourceMarker marker = new SourceMarker(new Buffer().writeUtf8(as + bs + cs + ds));
+    BufferedSource source = marker.source();
+
+    assertThat(source.readUtf8(10_000)).isEqualTo(as);
+    long pos10k = marker.mark(15_000);
+    assertThat(source.readUtf8(10_000)).isEqualTo(bs);
+    long pos20k = marker.mark(15_000);
+    assertThat(source.readUtf8(10_000)).isEqualTo(cs);
+    marker.reset(pos20k);
+    marker.reset(pos10k);
+    assertThat(source.readUtf8(30_000)).isEqualTo(bs + cs + ds);
+  }
+
+  private String repeat(char c, int count) {
```

Comment:
Java 11 has a `String.repeat(int count)` method so come september we can kill all these.

Reply:
Yay.